File: secator/tasks/prompt.py
import os
import signal
import click

from secator.decorators import task
from secator.output_types import Tag
from secator.runners import PythonRunner
from secator.rich import console
from secator.config import CONFIG


def _is_ci():
    """Check if running in CI environment."""
    return any(os.environ.get(var) for var in ('CI', 'CONTINUOUS_INTEGRATION', 'GITHUB_ACTIONS', 'GITLAB_CI', 'JENKINS_URL', 'BUILDKITE'))  # noqa: E501


def confirm_with_timeout(message, default=True, timeout=CONFIG.runners.prompt_timeout):
    """Prompt user with optional timeout.

    Args:
        message: The prompt message to display
        default: Default value if timeout occurs
        timeout: Timeout in seconds (0 to disable)

    Returns:
        User's response or default value on timeout
    """
    if timeout and timeout > 0:
        def timeout_handler(signum, frame):
            raise TimeoutError('Prompt timeout')

        old_handler = signal.signal(signal.SIGALRM, timeout_handler)
        signal.alarm(timeout)
        try:
            result = click.confirm(message, default=default)
        except (TimeoutError, KeyboardInterrupt):
            console.print(f'\n\[[bold red]PROMPT[/]] [bold red]Prompt timed out after {timeout}s, continuing...[/]')
            result = default
        finally:
            signal.alarm(0)
            signal.signal(signal.SIGALRM, old_handler)
        return result
    else:
        return click.confirm(message, default=default)


@task()
class prompt(PythonRunner):
    """Prompt the user."""
    output_types = [Tag]
    tags = ['network', 'recon']
    default_inputs = ''
    input_flag = None
    opts = {
        'yes': {
            'is_flag': True,
            'default': False,
            'short': 'y',
            'help': 'Auto-accept the prompt and forward all objects',
        },
        'message': {
            'type': str,
            'default': 'Validate {input}?',
            'help': 'The message to display to the user',
        },
    }

    def yielder(self):
        yes = self.run_opts.get('yes', False)
        in_ci = _is_ci()

        if len(self.inputs) == 0:
            return

        # Auto-accept in CI or if yes flag is set
        if yes or in_ci:
            console.print('\n\[[bold green]PROMPT[/]] [bold green]Auto-accepted all inputs.[/]')
        else:
            console.print('\n\[[bold red]PROMPT[/]] [bold red]Prompting user for validation ({timeout}s timeout)...[/]'.format(timeout=CONFIG.runners.prompt_timeout))  # noqa: E501

        for input in self.inputs:
            if yes or in_ci:
                result = True
            else:
                result = confirm_with_timeout(
                    self.run_opts['message'].format(input=input),
                    default=True,
                    timeout=CONFIG.runners.prompt_timeout
                )

            if result:
                yield Tag(
                    name='user_input',
                    match='auto' if (yes or in_ci) else 'prompt',
                    value=input,
                    category='info',
                )

File: rust/crates/secator-tasks/src/native/prompt.rs
//! prompt — interactive operator gate (Python `secator/tasks/prompt.py`).
//!
//! In a TTY, asks "Validate {input}?" (y/N, default = accept) per input and
//! emits `Tag(info, name=user_input, value=<input>, match=prompt)` for the
//! accepted ones. In CI / non-TTY / when `--yes` is set, auto-accepts every
//! input (emitting with `match=auto`). For non-interactive Rust parity right
//! now, we default to **auto-accept** unconditionally — the interactive read
//! path is wired but stays behind the `secator_prompt_interactive` env var
//! until the CLI plumbs `stdin().read_line`-with-timeout properly (Python's
//! `signal.alarm` doesn't translate cleanly).

use std::io::{self, IsTerminal, Write};

use secator_model::{OutputItem, Tag};
use secator_options::{OptSchema, OptSpec, OptType, RunOpts};
use secator_runner::{HookRegistry, NativeSpec, ValidatorRegistry};

pub static SPEC: NativeSpec = NativeSpec {
    name: "prompt",
    description: "Prompt the operator (auto-accepts in CI / non-TTY).",
    input_types: &[],
    output_types: &["tag"],
    tags: &["network", "recon"],
    run,
    hooks: HookRegistry::EMPTY,
    validators: ValidatorRegistry::EMPTY,
    schema: build_schema,
    install: secator_runner::InstallSpec::EMPTY,
    proxy_caps: secator_runner::ProxyCaps::NONE,
    encoding: "utf-8",
    ignore_return_code: false,
    requires_sudo: false,
    default_inputs: Some(""),
};

fn run(inputs: &[String], opts: &RunOpts) -> Vec<OutputItem> {
    if inputs.is_empty() {
        return Vec::new();
    }
    let yes = opts
        .get("yes")
        .map(|v| matches!(v.as_str(), "true" | "1" | "yes"))
        .unwrap_or(false);
    let in_ci = is_ci();
    let template = opts
        .get("message")
        .cloned()
        .unwrap_or_else(|| "Validate {input}?".into());
    let interactive_env = std::env::var("SECATOR_PROMPT_INTERACTIVE")
        .map(|v| v != "0" && !v.is_empty())
        .unwrap_or(false);
    let force_yes = yes || in_ci || !std::io::stdin().is_terminal() || !interactive_env;

    let mut out: Vec<OutputItem> = Vec::new();
    let match_label = if force_yes { "auto" } else { "prompt" };
    for input in inputs {
        let accepted = if force_yes {
            true
        } else {
            ask_user(&template.replace("{input}", input))
        };
        if accepted {
            out.push(OutputItem::Tag(Tag {
                name: "user_input".into(),
                match_: match_label.into(),
                value: input.clone(),
                category: "info".into(),
                ..Default::default()
            }));
        }
    }
    out
}

fn is_ci() -> bool {
    [
        "CI",
        "CONTINUOUS_INTEGRATION",
        "GITHUB_ACTIONS",
        "GITLAB_CI",
        "JENKINS_URL",
        "BUILDKITE",
    ]
    .iter()
    .any(|k| std::env::var(k).map(|v| !v.is_empty()).unwrap_or(false))
}

/// Interactive y/N read with a `CONFIG.runners.prompt_timeout`-second deadline.
/// Default is "yes" so both a bare Enter AND a timeout accept — matches
/// Python's `confirm_with_timeout(default=True, timeout=CONFIG.runners.prompt_timeout)`.
///
/// The read runs on a helper thread so the timeout can fire without leaving the
/// thread blocked on stdin forever. The thread will keep running after the
/// timeout and absorb the next line the operator types — acceptable for a
/// one-shot CLI process.
fn ask_user(message: &str) -> bool {
    let stderr = io::stderr();
    let mut h = stderr.lock();
    let _ = write!(h, "{message} [Y/n] ");
    let _ = h.flush();
    drop(h); // release before potentially blocking on the recv below

    let timeout_secs = secator_config::get().runners.prompt_timeout.max(0) as u64;
    let (tx, rx) = std::sync::mpsc::channel::<String>();
    std::thread::spawn(move || {
        let mut buf = String::new();
        if io::stdin().read_line(&mut buf).is_ok() {
            let _ = tx.send(buf);
        }
    });
    let recv = if timeout_secs == 0 {
        rx.recv().ok()
    } else {
        rx.recv_timeout(std::time::Duration::from_secs(timeout_secs)).ok()
    };
    match recv {
        Some(line) => {
            let s = line.trim().to_lowercase();
            s.is_empty() || s == "y" || s == "yes"
        }
        None => true, // timeout → default accept (Python parity)
    }
}

fn build_schema() -> OptSchema {
    let mut s = OptSchema { opt_prefix: "--", ..OptSchema::default() };
    s.opts = vec![
        OptSpec {
            name: "yes",
            ty: OptType::Bool,
            short: Some("y"),
            is_flag: true,
            default: None,
            help: "Auto-accept the prompt and forward all objects",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
        OptSpec {
            name: "message",
            ty: OptType::Str,
            short: None,
            is_flag: false,
            default: Some("Validate {input}?"),
            help: "Prompt message ({input} is interpolated)",
            internal: false,
            requires_sudo: false,
            shlex: true,
            pre_process: None,
            process: None,
        },
    ];
    s
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::BTreeMap;

    #[test]
    fn auto_accepts_in_ci_via_yes_flag() {
        let mut opts = BTreeMap::new();
        opts.insert("yes".into(), "true".into());
        let items = run(&["10.0.0.0/24".into(), "192.168.0.0/16".into()], &opts);
        let tags: Vec<&Tag> = items
            .iter()
            .filter_map(|i| match i { OutputItem::Tag(t) => Some(t), _ => None })
            .collect();
        assert_eq!(tags.len(), 2);
        for t in &tags {
            assert_eq!(t.name, "user_input");
            assert_eq!(t.match_, "auto");
            assert_eq!(t.category, "info");
        }
        assert_eq!(tags[0].value, "10.0.0.0/24");
        assert_eq!(tags[1].value, "192.168.0.0/16");
    }

    #[test]
    fn no_inputs_yields_nothing() {
        let items = run(&[], &BTreeMap::new());
        assert!(items.is_empty());
    }
}
